### PR TITLE
Improve farming

### PR DESCRIPTION
**Pull request by JoshdanG**
### Improve Farming Desire Calculation
1. Check for damage from enemy players instead of any player damage
- Fixes issue with (incorrectly) detecting self damage interrupting farming
2. Removed unhelpful conditions
- Check for channeling removed
- Check if current action is "idle" removed

### Improve Farming Behavior
1. Fix stacking location
- Fixes runtime error
2. Fix stacking without creeps
3. Mostly fixes leaving tree-heavy camps uncleared (while still farming)
4. Fixes leaving baby mud golems behind
5. Farming now uses 4 states: travelling, near, farming, stacking

### Misc Cleanup
1. Remove commented out and debugging / obsolete code
2. Add comments
3. Use more constants
4. Consolidate conditions required for farming
5. Move rattletrap specific farming code to separate function
6. Refactor / tweak camp refresh
- Move refresh to before ancient check so it always runs when it should
- Simplify 0:30 time check
- Change starting value of numCamp and add clarifying comment
7. New general use functions
- Unit.WasRecentlyDamagedByEnemy in MiraDota.lua
- utility.GetUnitsTowardsLocationGeneric in utility.lua

### CODE REVIEW

**Comment by JoshdanG** [comment]
### Reviewing
- Be sure to split up commits when reviewing.  The first 4 commits are refactoring and especially the rattletrap code move does not diff well.  
- Hiding whitespace is also helpful for reviewing the GetDesire() changes. 

### Feature Status
- Farming is currently only enabled for Alchemist and Bloodseeker -- the heroes that farming was previously enabled for.  
- Based on the previously configured desire levels, bots spend most of the time farming. This is probably not optimal, but is an interesting difference from typical bot behavior.
- Really doesn't work for ranged heroes
- There are plenty of improvements possible - like camp selection and balancing farming with other activities (especially pushing lanes)
- The camp clear detection is pretty dumb; it basically runs into each camp to be sure it is clear.

**Comment by adamqqqplay** [comment]
Thank you very much for your continued contribution.
I am a little busy at work recently, and code review will be slower. I'll find time to finish it though, thanks.

**Comment by ArcherWayne** [comment]
Thanks for your contribution. Farming has been a problem for this bot. 

I reviewed the code and I think the code are functional and logical. Can you try to test the GPM improvements for these bots?

If the bots can farm as fast as human players do, then they would be more challenging to play against.

**Comment by JoshdanG** [comment]
> Thank you very much for your continued contribution. I am a little busy at work recently, and code review will be slower. I'll find time to finish it though, thanks.

No worries.  Definitely no urgency from my end.

**Comment by JoshdanG** [comment]
> I reviewed the code and I think the code are functional and logical. Can you try to test the GPM improvements for these bots?
> 
> If the bots can farm as fast as human players do, then they would be more challenging to play against.

It's definitely not (good) human level, but it's pretty consistently higher than non-farming heroes. Some sample game scoreboards below (Alchemist: 690gpm, Bloodseeker: 632gpm, Bloodseeker: 775 gpm).  These games happen to all be wins, but the farming hero generally does not carry the game.

Alchemist
Alchemist_farm

Bloodseeker (bad game)
Bloodseeker_farm

Bloodseeker (good game)
Bloodseeker_farm2

File: mode_farm_generic.lua [hunk 1]
```diff
@@ -39,44 +36,52 @@ function GetDesire()
...
+	-- * Game is not mid only
+	-- * Have something to buy
+	-- * No enemies nearby, and none have damaged us recently
+	-- * No enemy heroes near our base
+	-- * We aren't trying to retreat or go to secret shop
+
+	if not (bot:IsAlive()
+			and campUtils.IsStrongJungler(bot)
+			and bot:GetLevel() >=6 and bot:GetLevel() < 30)
+		or IsHumanPlayerInTeam()
```
Comment (adamqqqplay) [inline mode_farm_generic.lua:57]: Is this condition necessary? I don't quite understand the difference when human players are present.

File: util/CampUtility.lua [hunk 2]
```diff
@@ -33,8 +12,12 @@ local jungler = {
 	--'npc_dota_hero_ursa'
 }
 
-function X.GetCampMoveToStack(id)
-	return CStackLoc[id]
+-- Get the location that bot should move to in order to stack the camp.
+--
+-- Currently works by just going a fixed distance away from camp location 
+-- in the bot's current dirction.
+function X.GetCampMoveToStack(bot, camp)
```
Comment (adamqqqplay) [inline util/CampUtility.lua:19]: This function is really helpful, thanks!

**Review by ArcherWayne — APPROVED** [review]

File: mode_farm_generic.lua (hunk 1 above)
Comment (JoshdanG) [reply mode_farm_generic.lua:57]: That logic was already there and I assume working as intended.  I'm guessing the intent is that the human player may want to farm, and the bot farming as well could interfere with that.  That check can be removed if unwanted; nothing else depends on it.

Comment (adamqqqplay) [reply mode_farm_generic.lua:57]: @JoshdanG Now we only have 2 heroes with farming mode enabled? If so, it is indeed possible to remove this condition, it affects human players less.

**Pull request merged**